numba/cpython/randomimpl.py

```diff
@@ -392,68 +507,188 @@ def get_num():
+
+
+@overload(np.random.randint)
+def np_randint_impl_1(high):
+    if isinstance(high, types.Integer):
+        return lambda high: np.random.randint(0, high)
+
+
+@overload(np.random.randint)
+def np_randint_impl_2(low, high):
+    if isinstance(low, types.Integer) and isinstance(high, types.Integer):
+        signed = max(low.signed, high.signed)
+        bitwidth = max(low.bitwidth, high.bitwidth)
+        int_ty = types.Integer.from_bitwidth(bitwidth, signed)
+        llvm_type = ir.IntType(bitwidth)
+
+        start_preprocessor = _randrange_preprocessor(bitwidth, low)
+        stop_preprocessor = _randrange_preprocessor(bitwidth, high)
+
+        @intrinsic
+        def _impl(typingcontext, low, high):
+            def codegen(context, builder, sig, args):
+                start, stop = args
+
+                start = start_preprocessor(builder, start, llvm_type)
+                stop = stop_preprocessor(builder, stop, llvm_type)
+                step = ir.Constant(llvm_type, 1)
+                return _randrange_impl(context, builder, start, stop, step,
+                                       llvm_type, signed, 'np')
+            return signature(int_ty, low, high), codegen
+        return lambda low, high: _impl(low, high)
+
+
+@overload(np.random.randint)
+def np_randint_impl_3(low, high, size):
+    if (isinstance(low, types.Integer) and isinstance(high, types.Integer) and
+       is_nonelike(size)):
+        return lambda low, high, size: np.random.randint(low, high)
+    if (isinstance(low, types.Integer) and isinstance(high, types.Integer) and
+       (isinstance(size, types.Integer) or (isinstance(size, types.UniTuple)
+                                            and isinstance(size.dtype,
+                                                           types.Integer)))):
+        bitwidth = max(low.bitwidth, high.bitwidth)
+        result_type = {8: np.int8,
+                       16: np.int16,
+                       32: np.int32,
+                       64: np.int64}[bitwidth]
+
+        def _impl(low, high, size):
+            out = np.empty(size, dtype=result_type)
+            out_flat = out.flat
+            for idx in range(out.size):
+                out_flat[idx] = np.random.randint(low, high)
+            return out
+        return _impl
+
+
+@overload(random.uniform)
+def uniform_impl0():
+    return lambda: random.uniform(0.0, 1.0)
+
+
+@overload(np.random.uniform)
+def np_uniform_impl0():
+    return lambda: np.random.uniform(0.0, 1.0)
+
+
+@overload(random.uniform)
+def uniform_impl1(low):
+    if isinstance(low, types.Number):
+        return lambda low: random.uniform(low, 1.0)
+
+
+@overload(np.random.uniform)
+def np_uniform_impl1(low):
+    if isinstance(low, types.Number):
+        return lambda low: np.random.uniform(low, 1.0)
+
+
+@overload(random.uniform)
+def uniform_impl2(low, high):
+    if isinstance(low, types.Number) and isinstance(high, types.Number):
```

Comment:
I'm slightly uncomfortable with accepting `Number` here (and throughout, it's a common occurrence in this PR) and then deferring the type check to e.g. `_double_preprocessor` because it's pushing the error further down the stack and relying on the presence of a call to something else to catch a problem. Would checking if the types are in `types.real_domain | types.integer_domain` be sufficient and ensure the error is raised from this overload opposed to the intrinsic call?

Reply:
Yeah, random complex numbers are probably not a good idea. Fixed.